src/details/ArborX_DetailsMutualReachabilityDistance.hpp

```diff
@@ -0,0 +1,93 @@
+/****************************************************************************
+ * Copyright (c) 2017-2021 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_DETAILS_MUTUAL_REACHABILITY_DISTANCE_HPP
+#define ARBORX_DETAILS_MUTUAL_REACHABILITY_DISTANCE_HPP
+
+#include <ArborX_AccessTraits.hpp>
+#include <ArborX_DetailsAlgorithms.hpp>
+#include <ArborX_DetailsHappyTreeFriends.hpp>
+#include <ArborX_DetailsKokkosExtMinMaxOperations.hpp>
+#include <ArborX_Predicates.hpp>
+
+namespace ArborX
+{
+namespace Details
+{
+
+template <class Primitives, class Distances>
+struct MaxDistance
+{
+  Primitives _primitives;
+  Distances _distances;
+  using Access = AccessTraits<Primitives, PrimitivesTag>;
+  using memory_space = typename Access::memory_space;
+  using size_type = typename memory_space::size_type;
+  template <class Predicate>
+  KOKKOS_FUNCTION void operator()(Predicate const &predicate, size_type i) const
+  {
+    size_type const j = getData(predicate);
+    using KokkosExt::max;
+    auto const distance_ij =
+        distance(Access::get(_primitives, i), Access::get(_primitives, j));
+    _distances(j) = max(_distances(j), distance_ij); // careful here
+  }
+};
+
+template <class Primitives>
+struct NearestK
+{
+  Primitives primitives;
+  int k;
+};
+
+} // namespace Details
+
+template <class Primitives>
+struct AccessTraits<Details::NearestK<Primitives>, PredicatesTag>
+{
+  using Access = AccessTraits<Primitives, PrimitivesTag>;
+  using memory_space = typename Access::memory_space;
+  using size_type = typename memory_space::size_type;
+  static KOKKOS_FUNCTION size_type size(Details::NearestK<Primitives> const &x)
+  {
+    return Access::size(x.primitives);
+  }
+  static KOKKOS_FUNCTION auto get(Details::NearestK<Primitives> const &x,
+                                  size_type i)
+  {
+    // NOTE account for self collision
+    return attach(nearest(Access::get(x.primitives, i), x.k + 1), i);
```

Comment:
```suggestion
    return attach(nearest(Access::get(x.primitives, i), x.k), i);
```
Core distance definition includes self collisions.

Reply:
The definition of core distance does not mean nearest k neighbors include self collision, it means `k = min_core_size + 1` which is dealt with at call site.

Reply:
While I understand where you are coming from, I think this is still wrong. For a simple reason it is not guaranteed that the set of primitives and predicates match. If they don't, then the provided answer is different from what a user asked, as it will return more neighbors than expected. You could say that it's going to only be used in HDBSCAN, but I think given its convenience it's likely be used more widely, leading to errors.

Reply:
I don't follow you argument.
I am saying that if I do
```C++
BVH bvh{exec_space, primitives};
bvh.query(exec_space, NearestK{primitives, k}, MaxDistance{primitives, distances});
```
it feels more natural to me that `k` excludes self-collisions

Reply:
Note that I would use half traversal from #485 here

Reply:
We still don't quite agree on whether self collision should be included but we identified the source of confusion.
Andrey meant using `NearestK{geometries}` in general where `geometries` would not necessarily be the same as the `primitives` argument that was used to build `bvh`, whereas my intention was only a facility to solve nearest problem where they are the same and definitely not to provide a general facility.  We haven't reached a resolution yet but at least I think we understand each other better.  We may decide to accept temporarily one convention or the other to make progress on this and resume the argument later.